sklearn/preprocessing/_polynomial.py

```diff
@@ -249,8 +249,12 @@ def transform(self, X):
                 to_stack.append(np.ones(shape=(n_samples, 1), dtype=X.dtype))
             to_stack.append(X)
             for deg in range(2, self.degree+1):
+                # Dimensionality of the expanded space can become very
+                # large so we cast X.shape[1] to int64 to avoid overflow
+                # when computing expanded_dimensionality
                 Xp_next = _csr_polynomial_expansion(X.data, X.indices,
-                                                    X.indptr, X.shape[1],
+                                                    X.indptr,
+                                                    np.int64(X.shape[1]),
```

Comment:
If X.indptr.dtype is int32, this would be downcasted anyways. We can change the signature of `_csr_polynomial_expansion` to always do the upcasting.

```python
def _csr_polynomial_expansion(...
                              np.int64_t d, INDEX_T interaction_only,
                              INDEX_T degree):
```